### DISCUSSION TITLE
TcpListener可以用在udp上面？还是说这个只是一个名字

### DISCUSSION THREAD

**Discussion by redoriental** [salvo-rs/salvo#298] [General]
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**

**Comment by chrislearn** [comment]
不能用在Udp上.
如果你是 Http3, 可以用 QuinnListener.

**Reply by redoriental** [reply]
但是你的webtransport的例子里面是TCP

**Reply by redoriental** [reply]
并且使用你的例子中的导入，并没有webtransport这个块

**Reply by redoriental** [reply]
tracing_subscriber::fmt().init();
    let cert = include_bytes!("../certs/cert.pem").to_vec();
    let key = include_bytes!("../certs/key.pem").to_vec();

    let router = Router::new().get(hello);
    let config = RustlsConfig::new(Keycert::new().cert(cert.as_slice()).key(key.as_slice()));
    let listener = TcpListener::new(("127.0.0.1", 5800)).rustls(config.clone());

    let acceptor = QuinnListener::new(config, ("127.0.0.1", 5800))
        .join(listener)
        .bind()
        .await;

    Server::new(acceptor).serve(router).await;这个地方这个Tcp监听是什么意思，h3不是udp嘛

**Reply by chrislearn** [reply]
webtransport 现在发布不了, 因为 h3 没法新版本, 且我需要自定义 WebTransportSession.
所以你只能使用现在的 git 作为 crate 地址.

**Comment by chrislearn** [comment]
其实浏览器并不知道你的服务器是支持 http3的, 所以第一次请求可以认为就是 http2 的请求, 当返回的头里面有支持 http3 的标识时, 下一次请求才会是 http3.
因此, 你必须开启 http2 的服务器.

**Reply by redoriental** [reply]
也就是说那里的tcp只是标识一下，实际上还是使用udp进行连接咯

**Reply by chrislearn** [reply]
这个 http2 服务器是真实有效的, 而且是必须创建 http3 才能正常工作的. 所以我不确定你说的"只是标识一下"跟我说的是一个意思.

**Reply by redoriental** [reply]
我懂你的意思了，连接的时候是2，传输的时候是3是吗，我刚刚观察了一下pid对端口的状态，发现tcp和udp都监听了，这和我之前使用的别的webtransport项目不一样，嘿嘿

**Reply by chrislearn** [reply]
这个应该是浏览器的一个行为. 如果你是 WebTransport, 因为这玩意必须是 http3, 所以, 我猜测不开始 http2, 浏览器应该也是会直接用 http3 建立连接.

**Reply by redoriental** [reply]
对，你也可以更改一下，这样更加体现了3的特征，第一次看见终于有中国人的webtransport项目，太牛逼了

**Reply by chrislearn** [reply]
呵呵, 我只是抄袭得快而已. 有问题也可以加微信群里讨论.

**Reply by redoriental** [reply]
有微信群嘛，我想交流一下，这种技术交流的太少了

**Reply by chrislearn** [reply]
加我微信 chrislearn
